### ISSUE TITLE
2D Wavelet Filtering

### ISSUE THREAD

**Issue by tommydino93** [Radiomics/pyradiomics#374]
Hi again!
I've been trying to apply the Wavelet filtering to my original 2D MRI cardiac image. I have been running the following code:

```
for decompositionImage, decompositionName, inputKwargs in imageoperations.getWaveletImage(cropped_image, cropped_mask):
    waveletFirstOrderFeaturs = firstorder.RadiomicsFirstOrder(decompositionImage, cropped_mask, **inputKwargs)
    waveletFirstOrderFeaturs.enableAllFeatures()
    waveletFirstOrderFeaturs.calculateFeatures()

    print('Calculated firstorder features with wavelet: ', decompositionName)
    for (key, val) in six.iteritems(waveletFirstOrderFeaturs.featureValues):
        waveletFeatureName = '%s_%s' % (str(decompositionName), key)
        print('  ', waveletFeatureName, ':', val)
```

and the shape features are calculated correctly. My question regards more the interpretation of these. As a matter of fact, 9 different shape_feature_groups are computed (LLL, LLH, LHL, LHH, HLL, HLH, HHL, HHH, LLL). My questions are:

1) How do I interpret the third letter of the filter (e.g. the H in LLH) since I only have a 2D image? Shouldn't I have just 4 filtered images (LL, LH, HL, HH)?

2) Why the LLL is computed twice?

3) Is there a reference article where the H and L filtering is well explained? From what I understood, the function applies a cascade filtering bank (e.g. HL means that first a high-pass filter and then a low-pass filter are applied to the image), but I'm not sure.

Thanks a lot in advance!

**Comment by tommydino93** [comment]
Hey @JoostJM !
By debugging the `_swt3` and the `getWaveletImage` functions (which I also commented a little bit in a pull request) I think I understood that the problem is with the `axes` variable. Since it is set to `axes = {2, 1, 0}` at the beginning of the function, then the filtering is performed in 3D. I have two alternatives, but I don't know if they are right:

1) I could manually modify the `axes` variable to e.g. `axes = {1,0}` or `axes = {0,1}` so as to not consider the z-axis. Does it make sense? Which one should I choose?

2) Could I use the `force2D` argument in this case?

Thanks a lot again for your time! :)

**Comment by tommydino93** [comment]
Hey @JoostJM!
In the end, we switched to 3D and not only a single 2D MR slice. I'll then restate [my question](https://github.com/Radiomics/pyradiomics/issues/374#issue-314553670):

Why the Wavelet-LLL feature group is computed twice? Now that I have a 3D dataset I simply copied the function in the `helloFeature` example, so I have:

```
for decompositionImage, decompositionName, inputKwargs in imageoperations.getWaveletImage(cropped_image, cropped_mask):
    # First Order
    waveletFirstOrderFeaturs = firstorder.RadiomicsFirstOrder(decompositionImage, cropped_mask, **inputKwargs)
    waveletFirstOrderFeaturs.enableAllFeatures()
    waveletFirstOrderFeaturs.calculateFeatures()
```

Thank you very much in advance for your amazing work!

**Comment by JoostJM** [comment]
@tommydino93 Thanks for the catch! The apparent double computation of LLL is indeed a bug, which I fixed in c8ceee2 (commited just now). It was not really computed twice, just returned twice.

As to your other questions:
1) The wavelet is computed regardless of your mask. So if you do not specify it as a forced 2D extraction, wavelet is computed in 3D. You can do this by indeed setting `force2D` to true. The axis defined in `force2Ddimension` is then removed from the axes variable (thereby calculating wavelet in 2D over the other 2 axes)
2) Bug, see above
3) That is indeed true. The best place to get more information on how exactly wavelets are computed is by taking a look at the PyWavelets package and documentation, which is the package we implement for the calculation of wavelets in PyRadiomics.
